tests/unit/steps/test_plan.py

```diff
@@ -1575,3 +1576,66 @@ def test_create_upgrade_plan_failed():
 
     with pytest.raises(Exception, match="test"):
         cou_plan._create_upgrade_group([app], "victoria", "test", False)
+
+
+@pytest.mark.parametrize(
+    "cli_force, empty_hypervisors, expected_logs",
+    [
+        (
+            True,
+            {},
+            [
+                (
+                    "cou.steps.plan",
+                    logging.INFO,
+                    "Selected all hypervisors: nova-compute/0, nova-compute/1, nova-compute/2",
+                )
+            ],
+        ),
+        (
+            False,
+            {1},
+            [
+                (
+                    "cou.steps.plan",
+                    logging.INFO,
+                    "Found non-empty hypervisors: nova-compute/0, nova-compute/2",
+                ),
+                ("cou.steps.plan", logging.INFO, "Selected hypervisors: nova-compute/1"),
+            ],
+        ),
+    ],
+)
+@pytest.mark.asyncio
+@patch("cou.steps.plan._get_nova_compute_units_and_machines")
+@patch("cou.steps.plan.get_empty_hypervisors")
+async def test_log_upgradable_hypervisors(
+    mock_empty_hypervisors,
+    mock_all_hypervisors,
+    cli_force,
+    empty_hypervisors,
+    expected_logs,
+    caplog,
+):
+    all_machines = [Machine(str(i), (), ()) for i in range(1, 4)]
+    empty_machines = [Machine(str(i + 1), (), ()) for i in empty_hypervisors]
+    all_units = [
+        Unit(
+            name=f"nova-compute/{i}",
+            machine=all_machines[i],
+            workload_version="21.0.0",
+        )
+        for i in range(len(all_machines))
+    ]
+    empty_units = [unit for unit in all_units if unit.machine in empty_machines]
+
+    analysis_result = MagicMock(spec_set=Analysis)()
+
+    mock_all_hypervisors.return_value = all_units, all_machines
+    mock_empty_hypervisors.return_value = empty_units, empty_machines
+
+    with caplog.at_level(logging.INFO):
+        await cou_plan._get_upgradable_hypervisors_machines(cli_force, analysis_result)
+
+    for expected_log in expected_logs:
+        assert expected_log in caplog.record_tuples
```

Comment:
I ended up rewriting the test and creating a separate config for it, for better flexibility. I think the test should focus on testing the logging logic, instead of the plan. Using a customized config can make it easier to control arguments such as cli_force.